src/Backups/BackupsWorker.cpp

```diff
@@ -864,6 +865,8 @@ OperationID BackupsWorker::startRestoring(const ASTPtr & query, ContextMutablePt
             /// process_list_element_holder is used to make an element in ProcessList live while RESTORE is working asynchronously.
             auto process_list_element = context_in_use->getProcessListElement();
 
+            /// we need to keep query context alive until RESTORE is done
+            auto query_context = CurrentThread::getQueryContext();
```

Comment:
But the `query_context` must be the same as `context_in_use` and it's kept alive already.

Reply:
couple of lines above you can see that if it's async we do this
```cpp
context_in_use = Context::createCopy(context);
```
so the initial context is not kept

Reply:
it's possible that the `if` needs to be `if (on_cluster)`
WDYT is better?

Reply:
> so the initial context is not kept

But do we need the initial context since we're using `context_in_use` to enqueue a DDL entry and not the initial context?

Reply:
Because the restore thread will attach to thread group created by the query and which contains the initial context.
We can also create a new QueryScope with `context_in_use` as context inside the restore thread.

Reply:
> it's possible that the if needs to be if (on_cluster) instead of if (restore_settings.async || on_cluster)
WDYT is better?

AFAIK contexts have not much of synchronization so we always copy a context if we pass it to another thread.

Reply:
> We can also create a new QueryScope with context_in_use as context inside the restore thread.

It's strange that we need that. Because `QueryScope::QueryScope()` mostly calls `CurrentThread::attachToGroup()` however ``CurrentThread::attachToGroup()` is called already - see the call of `scheduleFromThreadPool()` in `BackupsWorker::startRestoring()`.

Reply:
As I said, `ThreadGroup` stores `weak_ptr` to the context so `attachToGroup` won't extend the lifetime of it.
- run RESTORE async
- create query context and with QueryScope add it to thread group
- create restore thread and inside it:
      - attach to the group 
      - get the weak_ptr to context
- query finishes because it's async while the restore thread is running in background
- created query context is deleted and all weak_ptr to it become invalid

Reply:
BTW can we just set `internal = true` and pass this assertion? It seems it would be even correct logically.

Reply:
> We can also create a new QueryScope with context_in_use as context inside the restore thread.

Yes, it seems it would be better to create `QueryScope` before calling `doRestore()` on a separate thread.

Reply:
setting it to internal doesn't help, not sure why exactly 
I will try with `QueryScope`
Important we now know where the problem is, and I'm open to any solution

Reply:
also, do we want to do the same with backup inside `startMakingBackup`?